### PR TITLE
Zendesk: show alert if email is a8c.com/automattic.com

### LINKED ISSUE #921: Zendesk: show alert if email is a8c
**Describe the bug**
Ref. https://github.com/wordpress-mobile/WordPress-iOS/pull/11508

Zendesk does not allow agents to submit tickets, and displays a `Message failed to send` error upon attempt. We can't do anything about that, so we should show a warning when an `@automattic.com` or `@a8c.com` email address is detected:
- When the Support view is initially shown.
- When the user changes their email via `Contact Email`.

**To Reproduce**
1. Go to My store > Settings > Help & Support
2. If an email address is set to `a8c` an alert should be immediately shown (it doesn't)
3. If you set the contact email address row to `@a8c.com` and `@automattic.com` an alert should immediately be shown (it doesn't do anything right now) 

**Screenshots**
The alert that should appear:

**Expected behavior**
The alert should show at any point that a user has an `@a8c.com` or `@automattic.com` email address.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Mobile Environment**
Please include:
- Device: all
- iOS version: all
- WooCommerce iOS version: 1.7

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #921

Since developers cannot send support tickets in the app, there are two places in Help/Support where we want to show warning for developer with an @automattic.com or @a8c.com email address.

- When the Support view is initially shown (already implemented before this PR)
- When the user changes their email via Contact Email (implemented in this PR)

## Changes
- Updated `getUserInformationAndShowPrompt` completion block to the new completion type: `UserInformationCompletion = (_ success: Bool, _ email: String?) -> Void` so that email is available for upstream completion handler (`HelpAndSupportViewController`) to show warning if the email is a developer email
- Created `DeveloperEmailChecker` to check if an email has @automattic.com or @a8c.com domain + basic tests

## Testing
Prerequisites: if the contact email in Help & Support is currently set to a developer email, update email to a non-developer email like gmail by going to My store > Settings > Help & Support > Contact Email.
- Go to My store
- Go to Settings > Help & Support
- Set the contact email address row to @a8c.com and @automattic.com --> An alert should immediately be shown

Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WooCommerce/WooCommerceTests/Tools/DeveloperEmailCheckerTests.swift [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import XCTest
+@testable import WooCommerce
```
Comment (ctarda) [inline WooCommerce/WooCommerceTests/Tools/DeveloperEmailCheckerTests.swift]: Nice! 10_000 extra points for adding tests!

File: WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+struct DeveloperEmailChecker {
```
Comment (ctarda) [inline WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift]: Being a struct, it should not be that expensive to create an instance of this checker. On the one hand, the checker is a function, with a clear input and output, so it makes sense to declare it as static, but on the other hand we would not be able to pass an instance around if needed (which I don't know if it's an actual issue).

So, I'm curious, what made you decide declaring it as static?

File: WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    private static let developerEmails = ["@automattic.com", "@a8c.com"]
+
+    /// Checks if an email belongs to app developer (Automattic).
+    ///
+    static func isDeveloperEmail(email: String) -> Bool {
+        return (developerEmails.first { email.contains($0) }) != nil
```
Comment (ctarda) [inline WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift]: I haven't actually tested this, but should this work as well?

```
return developerEmails.contains(email)
```

I find that slightly easier to parse (of course, this is a personal preference, feel free to ignore it)

Comment (ctarda) [reply WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift]: Aaaaaand I just realised that my comment does not make sense, as we are checking the domain only 🤦‍♂ 🤷‍♂

File: WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift]: I generally prefer not using static whenever possible, like you said it's easier to pass around and DI, even though the function is purely functional. I picked static here after checking a few helper methods in the codebase that use static functional methods (like `CoreDataIterativeMigrator`, `SharingHelper`) and thought it's the convention. I can update to instance if we (Woo iOS devs @mindgraffiti) are all good with using instance methods for helpers like this case :)

Comment (mindgraffiti) [reply WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift]: Many of the Helpers you'll see in our codebase originally came from WPiOS, where (I believe) the convention was to make everything static. `CoreDataIterativeMigrator` is a port from WPiOS, but was converted to Swift, and `SharingHelper` also came from WPiOS.

If you check out our more recent Helpers, like `NSDecimalNumber+Helpers` or `Dictionary+Woo`, you'll see that we preferred instance methods. **However** all I can do is offer the historical context on our previous decisions. If you find areas to improve, by all means, we're open to suggestions :)

tl;dr: I agree with @ctarda and defer to his recommendation because it's not my area of expertise.

File: WooCommerce/WooCommerceTests/Tools/DeveloperEmailCheckerTests.swift (hunk 1 above)
Comment (mindgraffiti) [reply WooCommerce/WooCommerceTests/Tools/DeveloperEmailCheckerTests.swift]: Nice! I want to write unit tests too. I hope you can suggest things I should test :)

File: WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift [hunk 4]
```diff
@@ -156,7 +158,7 @@ class ZendeskManager: NSObject {
 
     /// Displays an alert allowing the user to change their Support email address.
     ///
-    func showSupportEmailPrompt(from controller: UIViewController, completion: @escaping (Bool) -> Void) {
+    func showSupportEmailPrompt(from controller: UIViewController, completion: @escaping UserInformationCompletion) {
```
Comment (mindgraffiti) [inline WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift]: nitpick: another convention we adopted was prefixing Completion handlers with `on`, so this could be `onUserInformationCompletion`

Comment (jaclync) [reply WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift]: updated!

File: WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/Tools/Developer/DeveloperEmailChecker.swift]: Thanks for the context and pointers! I updated this static method to instance method in https://github.com/woocommerce/woocommerce-ios/pull/1043/commits/35950ccbfb6fed5af89125306428ce7c574bb526 - I'd like to follow team convention on utility methods whether it's static or instance :)

**Comment by ctarda** [comment]
Thanks @jaclync ! :shipit: 

In general, once a pull request is approved, even if there are small nitpicks or suggestions from the reviewers, feel free to merge directly after updating (if the comments actually deserve an update) without requesting a second review. 

That way we can be slightly more efficient.
